### PR TITLE
feat(data frame): Add `.data_view_rows()`, `.sort()`, `.filter()`, `.update_sort()`, and `.update_filter()`; `cell_selection()` no longer returns `None`

### LINKED ISSUE #1345: Recent update to Shiny 0.9.0  broke DataTable input
Hello,

The recent update to Shiny 0.9.0 broke the DataTable input, which was one of the updates in this new version.
At this point `input.<tableID>_selected_rows()` stopped working and is no longer updating any reactive function. 
You can check [your own demo example](https://shinylive.io/py/editor/#code=NobwRAdghgtgpmAXAAjFADugdOgnmAGlQGMB7CAFzkqVQDMAnUmZdKAG3gfWoHMBXAJYQAzskEx0pBhWTtSUACYB9HhAHCRAHQiNmyEQAthucZOmyG1RXAY69LIyaxwAHuisixEqTPER0fgoiIR0dNQ1RZABeOQUVCKFRAAoASjCIISxDACZkrTAABQ4uZEK+JO0wdIgMgAErCBsGLFCmuDpkJgB3ETTEHWQhrtJemP9AihwKzWVFOmUROHY4YioVHr7UoYBiZAAefcHhpZW1uEVxgqICrAArUmFkkQoGZMFtumlxfwMLC+Sm1S20EnU2yGWS2QBQAcuQ4AVjkMrBR+AwIMg6AUAEqjMSnVbrFAgAnnRQAX0RtWpDWstiwiigFCgykYsAR7U6iVm836SK6cFR6IF7RaABEmVAACpQABGK2S3NERB6i2WhME5GUMFINmiBREwl4KwK22Qe0OOkIqAouB4tDAVFcFDA5IAukA) provided in the [documentation](https://shiny.posit.co/py/components/outputs/datatable/index.html#select-rows) to see it is broken.

I hope this can be patched soon :) Downgrading to version 0.8.1 reverts the issue.

Thanks!

### LINKED ISSUE #1351: Exception when using `.data_view()` with cleared table
Hi, discovered some exception thrown when using `.data_view()` with table that has been recently cleared.

Let me show an example as it clearly demonstrate the issue.

```py
from shiny import App, ui, reactive, render
import pandas as pd

app_ui = ui.page_fluid(
    ui.input_action_button("add_row", "Add row"),
    ui.input_action_button("clear_table", "Clear table"),
    ui.output_text("number_of_selected_rows"),
    ui.output_data_frame("df1"),
)


def server(input):
    df = reactive.Value(pd.DataFrame(columns=["A", "B"]))

    @render.data_frame
    def df1():
        return render.DataGrid(df(), selection_mode="rows")

    @reactive.effect
    @reactive.event(input.add_row)
    def _():
        old_df = df()
        new_df = pd.concat([old_df, pd.DataFrame([[1, 2]], columns=["A", "B"])])
        df.set(new_df)

    @render.text
    def number_of_selected_rows():
        df_selected = df1.data_view(selected=True)
        return f"Selected rows: {len(df_selected)}"

    @reactive.effect
    @reactive.event(input.clear_table)
    def _():
        df.set(pd.DataFrame(columns=["A", "B"]))


app = App(app_ui, server)
```

When you add a row, click on it and click the clear button you get:

<details>

<summary>Exception/Error that I got:</summary>

```
Traceback (most recent call last):
  File "/Users/pasza/miniconda3/envs/shiny/lib/python3.12/site-packages/pandas/core/indexing.py", line 1714, in _get_list_axis
    return self.obj._take_with_is_copy(key, axis=axis)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pasza/miniconda3/envs/shiny/lib/python3.12/site-packages/pandas/core/generic.py", line 4153, in _take_with_is_copy
    result = self.take(indices=indices, axis=axis)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pasza/miniconda3/envs/shiny/lib/python3.12/site-packages/pandas/core/generic.py", line 4133, in take
    new_data = self._mgr.take(
               ^^^^^^^^^^^^^^^
  File "/Users/pasza/miniconda3/envs/shiny/lib/python3.12/site-packages/pandas/core/internals/managers.py", line 891, in take
    indexer = maybe_convert_indices(indexer, n, verify=verify)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pasza/miniconda3/envs/shiny/lib/python3.12/site-packages/pandas/core/indexers/utils.py", line 282, in maybe_convert_indices
    raise IndexError("indices are out-of-bounds")
IndexError: indices are out-of-bounds

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/pasza/miniconda3/envs/shiny/lib/python3.12/site-packages/shiny/session/_session.py", line 1220, in output_obs
    value = await renderer.render()
            ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pasza/miniconda3/envs/shiny/lib/python3.12/site-packages/shiny/render/renderer/_renderer.py", line 255, in render
    value = await self.fn()
            ^^^^^^^^^^^^^^^
  File "/Users/pasza/miniconda3/envs/shiny/lib/python3.12/site-packages/shiny/render/renderer/_renderer.py", line 349, in __call__
    return await self._fn()
           ^^^^^^^^^^^^^^^^
  File "/Users/pasza/miniconda3/envs/shiny/lib/python3.12/site-packages/shiny/_utils.py", line 263, in fn_async
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/pasza/Projects/InnovationLead/playground/app_df.py", line 28, in number_of_selected_rows
    df_selected = df1.data_view(selected=True)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pasza/miniconda3/envs/shiny/lib/python3.12/site-packages/shiny/render/_data_frame.py", line 285, in data_view
    return self._data_view_selected()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pasza/miniconda3/envs/shiny/lib/python3.12/site-packages/shiny/reactive/_reactives.py", line 283, in __call__
    return _utils.run_coro_sync(self.get_value())
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pasza/miniconda3/envs/shiny/lib/python3.12/site-packages/shiny/_utils.py", line 378, in run_coro_sync
    coro.send(None)
  File "/Users/pasza/miniconda3/envs/shiny/lib/python3.12/site-packages/shiny/reactive/_reactives.py", line 293, in get_value
    raise self._error[0]
  File "/Users/pasza/miniconda3/envs/shiny/lib/python3.12/site-packages/shiny/reactive/_reactives.py", line 328, in _run_func
    self._value.append(await self._fn())
                       ^^^^^^^^^^^^^^^^
  File "/Users/pasza/miniconda3/envs/shiny/lib/python3.12/site-packages/shiny/_utils.py", line 263, in fn_async
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/pasza/miniconda3/envs/shiny/lib/python3.12/site-packages/shiny/render/_data_frame.py", line 524, in self__data_view_selected
    return _subset_data_view(selected=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pasza/miniconda3/envs/shiny/lib/python3.12/site-packages/shiny/render/_data_frame.py", line 515, in _subset_data_view
    return data.iloc[data_view_indices]
           ~~~~~~~~~^^^^^^^^^^^^^^^^^^^
  File "/Users/pasza/miniconda3/envs/shiny/lib/python3.12/site-packages/pandas/core/indexing.py", line 1191, in __getitem__
    return self._getitem_axis(maybe_callable, axis=axis)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pasza/miniconda3/envs/shiny/lib/python3.12/site-packages/pandas/core/indexing.py", line 1743, in _getitem_axis
    return self._get_list_axis(key, axis=axis)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/pasza/miniconda3/envs/shiny/lib/python3.12/site-packages/pandas/core/indexing.py", line 1717, in _get_list_axis
    raise IndexError("positional indexers are out-of-bounds") from err
IndexError: positional indexers are out-of-bounds
```

<details>

I think something's wrong inside shiny, am I right?
Initially, I thought `df1.data_view(selected=False)` is just `None` but the error is weird, but it's not the case.

### LINKED ISSUE #1378: Add `.data_view_info()` to `@render.data_frame` renderer
> [I'd like to have access to the selected rows]

Good news...

With a combination of 
https://github.com/posit-dev/py-shiny/blob/3b742a2240774521a16c688e9d73c3a80206ce43/shiny/render/_data_frame.py#L405-L412 and https://github.com/posit-dev/py-shiny/blob/3b742a2240774521a16c688e9d73c3a80206ce43/shiny/render/_data_frame.py#L492-L502, maybe we expose them via `.data_view_indices(*,  selected: bool = False)`? (Similar to `.data_view(*, selected: bool = False)`)

It could return a `{rows: list[int], cols: list[int]}` object containing all rows/cols being used (given `selected` value). We could add a `type` to describe how the values were selected, similar to the `CellSelection` type https://github.com/posit-dev/py-shiny/blob/3b742a2240774521a16c688e9d73c3a80206ce43/shiny/render/_data_frame_utils/_selection.py#L162-L185 . 

For the return value, I still don't know if `rows` and `cols` should always exist (containing all values). If they always exist, it is easier to code as a user. If they do not exist, it is much clearer as to what is being subsetted but more annoying to integrate with as a user.  I'm currently leaning towards always including `rows` and `cols` containing all values being used. Ex: multi row selection would return `{type: "row", rows: [1,3,5], cols: [0,1,2,3,4,5]}` assuming we selected rows 1, 3, and 5 and there were 6 columns.

```python
class DataViewIndices(TypedDict):
    type: Literal["row", "col", "rect", "none"]
    rows: ListOrTuple[int]
    cols: ListOrTuple[int]

# class data_frame...
def data_view_indices(self, *, selected: bool = False) -> DataViewIndices:
    ...
```

-----------------------

@pstorozenko Glad you like `.data_view()`!

> It's not a problem for now, but if one day we shiny switches to some dataframe representation without creature like `pd.Index` it may be.

Can I get a copy of your crystal ball? 😆 

I'd love to move to something like https://github.com/data-apis/dataframe-api-compat , however to [get typing support](https://github.com/data-apis/dataframe-api-compat/blob/main/contributing.md#type-checking), their typing specification is defined deep within another repo: https://github.com/data-apis/dataframe-api/tree/main/spec/API_specification/dataframe_api .  

I do not understand why the types being used within `dataframe-api-compat` are not shipped with the package. /rant

_Originally posted by @schloerke in https://github.com/posit-dev/py-shiny/issues/1345#issuecomment-2101178383_

### PR DESCRIPTION
**Pull request by schloerke**
Fixes #1351
Fixes #1345
Fixes #1378 

Related to https://github.com/posit-dev/py-shiny/issues/1345#issue-2271580277
Related discussion https://github.com/posit-dev/py-shiny/issues/1345#issuecomment-2101178383 with @pstorozenko

Final solution:

~~Expose `.data_view_info()` reactive value which returns `DataViewInfo` which is defined as a list of data view information.~~

Expose individual reactive values for the user to access on their own:

* `.data_view_rows: Calc_[tuple[int]]`
* `.sort: Calc_[ListOrTuple[ColumnSort]]`
* `.filter: Calc_[ListOrTuple[ColumnFilter]]`
* `.update_sort(self, sort: ListOrTuple[ColumnSort | int] | ColumnSort | int | None)`
	* `int` values are upgraded to `{col: <INT>, desc: bool}`; `desc` defaults to True if column `<INT>` is number like, otherwise `False`.
* `.update_filter(self, sort: ListOrTuple[ColumnFilter] | None)`

```python
class ColumnSort(TypedDict):
    col: int
    desc: bool

class ColumnFilterStr(TypedDict):
    col: int
    value: str

class ColumnFilterNumber(TypedDict):
    col: int
    # Similar to...
    value: tuple[float, float] | tuple[float, None] | tuple[None, float]

ColumnFilter = Union[ColumnFilterStr, ColumnFilterNumber]
```

----------------------

----------------------

From #1376

Previous return type for `.input_cell_selection()`:

```python
class CellSelectionNone(TypedDict):
    type: Literal["none"]

class CellSelectionRow(TypedDict):
    type: Literal["row"]
    rows: ListOrTuple[int]

class CellSelectionCol(TypedDict):
    type: Literal["col"]
    cols: ListOrTuple[int]

class CellSelectionRect(TypedDict):
    type: Literal["rect"]
    # Attempt to type the list of size 2, but it is not type enforced
    rows: tuple[int, int] | Annotated[list[int], 2]
    cols: tuple[int, int] | Annotated[list[int], 2]

CellSelection = Union[
    CellSelectionRow,
    CellSelectionCol,
    CellSelectionRect,
    CellSelectionNone,
]

def .cell_selection() -> CellSelection | None:
```

New return type for `.cell_selection()`:

```python
class CellSelection(TypedDict):
    type: Literal["none", "row", "col", "rect"]
    rows: ListOrTuple[int]
    cols: ListOrTuple[int]

def .cell_selection() -> CellSelection:
```

* If `type = "none"`, `rows` and `cols` are empty tuples. 
* If `type = "row"`, `cols` is all column numbers of the data.
* If `type = "col"`, `rows` is all row numbers of the data view rows
* If `type = "rect"`, all rows and cols are provided (no longer just min/max values). 

* If `.cell_selection()` receives an empty selection (e.g. `type="row"` and `rows=[]`, or `type="none"`), then the return value will be `None` as no selection is currently taking place. 

Previous user interaction: 

```python
@render.data_frame
def grid(): ...

selected_rows = (grid.cell_selection() or {}).get("rows", ())
```

New user interaction:

```python
selected_rows = grid.cell_selection()["rows"]
```

### CODE REVIEW

File: shiny/render/_data_frame.py [hunk 1]
```diff
@@ -78,6 +76,18 @@ class ColumnFilterNumber(TypedDict):
     value: tuple[float, float]
```
Comment (schloerke) [inline shiny/render/_data_frame.py:76]:
```suggestion
    value: tuple[float | None, float | None]
```

File: shiny/render/_data_frame.py [hunk 2]
```diff
@@ -275,6 +285,18 @@ def data_view(self, *, selected: bool = False) -> pd.DataFrame:
         else:
             return self._data_view_all()
 
+    data_view_info: reactive.Calc_[DataViewInfo]
```
Comment (wch) [inline shiny/render/_data_frame.py:288]: Add public-facing examples that use this.

For example: an app where there is an interactive table, and below it, a code output that prints out:
- The values of `sort` and `filter`
- A data frame using `rows`, which reflects exactly the sorted and filtered data that the user is seeing in the interactive table, and also has an extra column indicating whether that row is selected.
- Also add a second code output that just prints the value of `.data_view(selected=TRUE)`.

File: CHANGELOG.md [hunk 3]
```diff
@@ -9,8 +9,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Breaking Changes
 
+* `@render.data_frame`'s input value `input.<ID>_data_view_indices` has been renamed to `input.<ID>_data_view_rows` for consistent naming. Please use `input.<ID>_data_view_rows` and consider `input.<ID>_data_view_indices` deprecated. (#1374)
+
+* Restored `@render.data_frame`'s (prematurely removed in v0.9.0) input value `input.<ID>_selected_rows()`. Please use `<ID>.input_cell_selection()["rows"]` and consider `input.<ID>_selected_rows()` deprecated. (#1345)
```
Comment (wch) [inline CHANGELOG.md:14]: Move this change to a separate PR which we can merge right away.

File: shiny/render/_data_frame.py [hunk 4]
```diff
@@ -380,36 +412,53 @@ def self_input_cell_selection() -> CellSelection | None:
 
         self.input_cell_selection = self_input_cell_selection
 
-        # # Array of sorted column information
-        # # TODO-barret-render.data_frame; Expose and update column sorting
-        # # Do not expose until update methods are provided
```
Comment (wch) [inline shiny/render/_data_frame.py:385]: Add `update` method for sort.

File: shiny/render/_data_frame.py [hunk 5]
```diff
@@ -380,36 +412,53 @@ def self_input_cell_selection() -> CellSelection | None:
...
+
+        @reactive.calc
+        def self_data_view_info() -> DataViewInfo:
+
+            cell_selection = self.input_cell_selection()
+            selected_rows = tuple(
+                cell_selection["rows"]
+                if cell_selection is not None and "rows" in cell_selection
+                else ()
+            )
```
Comment (wch) [inline shiny/render/_data_frame.py:439]: Double check that this is ordered the same as `rows`.

File: shiny/render/_data_frame.py [hunk 6]
```diff
@@ -2,6 +2,9 @@
 
 import warnings
 
+# TODO-barret; Should `.input_cell_selection()` ever return None? Is that value even helpful? Empty lists would be much more user friendly.
+# TODO-barret-render.data_frame; Add method `update_sort()`
```
Comment (schloerke) [inline shiny/render/_data_frame.py:6]: Add method `update_sort()`

File: shiny/render/_data_frame.py [hunk 7]
```diff
@@ -2,6 +2,9 @@
...
+# TODO-barret-render.data_frame; Add method `update_filter()`
```
Comment (schloerke) [inline shiny/render/_data_frame.py:7]: Add method `update_filter()`

File: CHANGELOG.md (hunk 3 above)
Comment (schloerke) [reply CHANGELOG.md:18]: #1377

**Comment by jcheng5** [comment]
After Friday discussion: drop `DataViewInfo`, instead add rows as another accessor on the renderer. Also, input_cell_selection rows should be ordered the way the user sees it.
